driver.go

```diff
@@ -81,7 +81,8 @@ func (d MySQLDriver) Open(dsn string) (driver.Conn, error) {
 		}
 	}
 
-	mc.buf = newBuffer(mc.netConn)
+	// Make the buffer get the context from mysqlConn.
+	mc.buf = newBufferContext(mc.netConn, &ContextWrapper{mc: mc})
```

Comment:
> if Pinger.Ping() returns ErrBadConn. In other words, Golang database/sql/sql.go has already provided the mechanism to close a driver.Conn(mysql.mysqlConn), and indirectly destroy its underlying resource.

It's true only in Ping.  We can't use ErrBadConn for QueryContext() or other APIs to close connection from caller soon.
I want to use same pattern as possible for Ping, QueryContext() and other APIs.

> https://golang.org/pkg/database/sql/driver/
> To prevent duplicate operations, ErrBadConn should NOT be returned if there's a possibility that the database server might have performed the operation. Even if the server sends back an error, you shouldn't return ErrBadConn.